src/tools/grpc_server.cc

```diff
@@ -235,8 +290,58 @@ Status CodeSearchImpl::Search(ServerContext* context, const ::Query* request, ::
         return Status(StatusCode::INVALID_ARGUMENT, "Parse error");
     }
 
+    string line_pat = q.line_pat->pattern();
+
+    /* Patterns like "User.*Info" and "p\d+" might be genuine attempts
+       to match tags, so we cannot too quickly dismiss odd-looking REs
+       as justifying our skipping the phases of a tags search.  But we
+       can at least special-case a few characters that would not ever
+       appear in a pattern that matches tags.
+       TODO(brandon-rhodes): make this more sophisticated. */
+    bool might_match_tags =
+        // Characters that can't appear in an RE that matches a tag.
+        line_pat.find_first_of(" !\"#%&',-/;<=>@`") == string::npos
+        // If the user anchored the RE, then we can only run it against
+        // whole lines from the corpus, never against tags.
+        && line_pat.front() != '^'
+        && line_pat.back() != '$'
+        ;
+
     match_stats stats;
-    if (q.tags_pat == NULL) {
+    if (q.tags_pat == NULL && tagdata_ && might_match_tags) {
+        string regex;
+        int32_t max_matches = q.max_matches;  // remember original value
+
+        add_match::line_set unique_lines;
+        add_match cb(&unique_lines, response);
+
+        /* To surface the most important matches first, start with tags.
+           First pass: is the pattern an exact match for any tags? */
+        regex = "^" + line_pat + "$";
+        q.line_pat.reset(new RE2(regex, q.line_pat->options()));
+        run_tags_search(q, tagdata_, cb, tagmatch_, stats);
+
+        q.max_matches = max_matches - cb.match_count();
+        if (q.max_matches > 0) {
```

Comment:
Let's extract this whole inner bit into a function, and then instead of these continually-nesting `if (q.max_matches > 0)` clauses, we can  `if (q.max_matches <= 0) return` and keep the indentation under control.

Reply:
Oh, nice, I was indeed bothered by the creeping indentation and that will solve it. It wasn't clear to my eyes that the argument list wouldn't need to be horrendously long to provide all the context used here, but I'll see what it looks like.